File: frontend/src/services/retentionService.ts
import { api } from '@/utils/api';

export interface DataRetentionPolicy {
  id: string;
  policy_name: string;
  policy_code: string;
  table_name: string;
  data_category: string;
  retention_period_days: number;
  retention_action: string;
  description?: string;
  legal_basis?: string;
  conditions?: Record<string, any>;
  status: string;
  last_executed?: string;
  next_execution?: string;
  created_at: string;
  updated_at: string;
}

export interface DataRetentionPolicyRequest {
  policy_name: string;
  policy_code: string;
  table_name: string;
  data_category: string;
  retention_period_days: number;
  retention_action?: string;
  description?: string;
  legal_basis?: string;
  conditions?: Record<string, any>;
}

export interface RetentionExecution {
  id: string;
  policy_id: string;
  execution_date: string;
  records_processed: number;
  records_deleted: number;
  records_archived: number;
  records_anonymized: number;
  status: string;
  error_message?: string;
  execution_time_seconds?: number;
}

export interface RetentionDashboard {
  total_policies: number;
  active_policies: number;
  policies_due: number;
  recent_executions: RetentionExecution[];
  total_records_processed: number;
  storage_saved_mb: number;
}

/**
 * Data Retention Service
 * Provides methods to interact with the data retention API endpoints
 */
export default {
  /**
   * Create a data retention policy
   * @param policyRequest - Policy creation request
   * @returns Promise with the created policy
   */
  async createPolicy(policyRequest: DataRetentionPolicyRequest) {
    return api.post('/retention/policies', policyRequest);
  },

  /**
   * List data retention policies
   * @param activeOnly - Whether to show only active policies
   * @returns Promise with the list of policies
   */
  async listPolicies(activeOnly: boolean = true) {
    return api.get(`/retention/policies?active_only=${activeOnly}`);
  },

  /**
   * Execute a specific retention policy
   * @param policyId - Policy ID to execute
   * @returns Promise with execution result
   */
  async executePolicy(policyId: string) {
    return api.post(`/retention/policies/${policyId}/execute`);
  },

  /**
   * Execute all due retention policies
   * @returns Promise with execution results
   */
  async executeAllPolicies() {
    return api.post('/retention/execute-all');
  },

  /**
   * Get retention execution history
   * @param policyId - Optional policy ID filter
   * @param limit - Maximum number of records
   * @returns Promise with execution history
   */
  async getExecutionHistory(policyId?: string, limit: number = 100) {
    const params = new URLSearchParams();
    if (policyId) params.append('policy_id', policyId);
    params.append('limit', limit.toString());

    return api.get(`/retention/executions?${params.toString()}`);
  },

  /**
   * Initialize default retention policies
   * @returns Promise with initialization result
   */
  async initializeDefaultPolicies() {
    return api.post('/retention/initialize');
  },

  /**
   * Get retention dashboard data
   * @returns Promise with dashboard data
   */
  async getDashboard() {
    return api.get('/retention/dashboard');
  },

  /**
   * Utility functions for data retention
   */
  utils: {
    /**
     * Format retention action for display
     */
    formatRetentionAction(action: string): string {
      const actionMap: Record<string, string> = {
        'delete': 'Delete',
        'archive': 'Archive',
        'anonymize': 'Anonymize'
      };
      
      return actionMap[action] || action.charAt(0).toUpperCase() + action.slice(1);
    },

    /**
     * Get action color for UI
     */
    getActionColor(action: string): string {
      const colorMap: Record<string, string> = {
        'delete': 'error',
        'archive': 'warning',
        'anonymize': 'info'
      };
      
      return colorMap[action] || 'grey';
    },

    /**
     * Format retention period for display
     */
    formatRetentionPeriod(days: number): string {
      if (days < 30) {
        return `${days} days`;
      } else if (days < 365) {
        const months = Math.round(days / 30);
        return `${months} month${months > 1 ? 's' : ''}`;
      } else {
        const years = Math.round(days / 365);
        return `${years} year${years > 1 ? 's' : ''}`;
      }
    },

    /**
     * Get status color for UI
     */
    getStatusColor(status: string): string {
      const colorMap: Record<string, string> = {
        'active': 'success',
        'inactive': 'warning',
        'suspended': 'error',
        'completed': 'success',
        'failed': 'error',
        'running': 'info'
      };
      
      return colorMap[status] || 'grey';
    },

    /**
     * Format data category for display
     */
    formatDataCategory(category: string): string {
      return category.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
    },

    /**
     * Calculate next execution time
     */
    getNextExecutionTime(lastExecuted?: string, periodDays: number = 1): Date {
      const base = lastExecuted ? new Date(lastExecuted) : new Date();
      return new Date(base.getTime() + (periodDays * 24 * 60 * 60 * 1000));
    },

    /**
     * Check if policy is due for execution
     */
    isPolicyDue(nextExecution?: string): boolean {
      if (!nextExecution) return false;
      return new Date(nextExecution) <= new Date();
    },

    /**
     * Format execution summary
     */
    formatExecutionSummary(execution: RetentionExecution): string {
      const { records_processed, records_deleted, records_archived, records_anonymized } = execution;
      
      if (records_deleted > 0) {
        return `Deleted ${records_deleted} of ${records_processed} records`;
      } else if (records_archived > 0) {
        return `Archived ${records_archived} of ${records_processed} records`;
      } else if (records_anonymized > 0) {
        return `Anonymized ${records_anonymized} of ${records_processed} records`;
      } else {
        return `Processed ${records_processed} records`;
      }
    },

    /**
     * Validate retention period
     */
    validateRetentionPeriod(days: number, dataCategory: string): { valid: boolean; message?: string } {
      if (days < 1) {
        return { valid: false, message: 'Retention period must be at least 1 day' };
      }
      
      // Category-specific validation
      const categoryLimits: Record<string, { min: number; max: number }> = {
        'audit': { min: 2555, max: 3650 }, // 7-10 years for audit data
        'financial': { min: 2555, max: 3650 }, // 7-10 years for financial data
        'session': { min: 1, max: 90 }, // 1-90 days for session data
        'security': { min: 30, max: 365 }, // 30 days to 1 year for security data
        'personal': { min: 30, max: 2555 } // 30 days to 7 years for personal data
      };
      
      const limits = categoryLimits[dataCategory];
      if (limits) {
        if (days < limits.min) {
          return { 
            valid: false, 
            message: `${this.formatDataCategory(dataCategory)} data must be retained for at least ${this.formatRetentionPeriod(limits.min)}` 
          };
        }
        if (days > limits.max) {
          return { 
            valid: false, 
            message: `${this.formatDataCategory(dataCategory)} data retention cannot exceed ${this.formatRetentionPeriod(limits.max)}` 
          };
        }
      }
      
      return { valid: true };
    },

    /**
     * Get recommended retention periods by data category
     */
    getRecommendedRetention(dataCategory: string): number[] {
      const recommendations: Record<string, number[]> = {
        'audit': [2555, 3650], // 7, 10 years
        'financial': [2555, 3650], // 7, 10 years
        'session': [7, 30, 90], // 1 week, 1 month, 3 months
        'security': [90, 180, 365], // 3 months, 6 months, 1 year
        'personal': [365, 1095, 2555], // 1 year, 3 years, 7 years
        'operational': [30, 90, 365] // 1 month, 3 months, 1 year
      };
      
      return recommendations[dataCategory] || [30, 90, 365];
    }
  }
};

File: backend/app/services/retention/retention_service.py
"""
Data retention service for managing data lifecycle.
"""
from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any

from sqlalchemy import text, and_, desc
from sqlalchemy.orm import Session
from uuid import UUID
import time

from app.models.data_retention import DataRetentionPolicy, RetentionExecution, RetentionAction, RetentionStatus





class DataRetentionService:
    """Service for managing data retention policies."""
    
    def __init__(self, db: Session):
        """  Init  ."""
        self.db = db
    
    def create_policy(self, policy_data: Dict[str, Any], created_by: UUID) -> DataRetentionPolicy:
        """Create Policy."""
        """Create a data retention policy."""
        policy = DataRetentionPolicy(
            policy_name=policy_data['policy_name'],
            policy_code=policy_data['policy_code'],
            table_name=policy_data['table_name'],
            data_category=policy_data['data_category'],
            retention_period_days=policy_data['retention_period_days'],
            retention_action=policy_data.get('retention_action', RetentionAction.DELETE),
            description=policy_data.get('description'),
            legal_basis=policy_data.get('legal_basis'),
            conditions=policy_data.get('conditions'),
            status=RetentionStatus.ACTIVE,
            next_execution=datetime.utcnow() + timedelta(days=1),
            created_by=created_by,
            updated_by=created_by
        )
        
        self.db.add(policy)
        self.db.commit()
        self.db.refresh(policy)
        
        return policy
    
    def execute_policy(self, policy_id: UUID) -> RetentionExecution:
        """Execute Policy."""
        """Execute a data retention policy."""
        policy = self.get_policy(policy_id)
        if not policy:
            raise ValueError(f"Policy {policy_id} not found")
        
        if policy.status != RetentionStatus.ACTIVE:
            raise ValueError(f"Policy {policy.policy_code} is not active")
        
        start_time = time.time()
        execution = RetentionExecution(
            policy_id=policy_id,
            status="running"
        )
        
        self.db.add(execution)
        self.db.flush()
        
        try:
            cutoff_date = datetime.utcnow() - timedelta(days=policy.retention_period_days)
            
            if policy.retention_action == RetentionAction.DELETE:
                result = self._delete_records(policy, cutoff_date)
            elif policy.retention_action == RetentionAction.ARCHIVE:
                result = self._archive_records(policy, cutoff_date)
            elif policy.retention_action == RetentionAction.ANONYMIZE:
                result = self._anonymize_records(policy, cutoff_date)
            else:
                raise ValueError(f"Unknown retention action: {policy.retention_action}")
            
            execution.records_processed = result.get('processed', 0)
            execution.records_deleted = result.get('deleted', 0)
            execution.records_archived = result.get('archived', 0)
            execution.records_anonymized = result.get('anonymized', 0)
            execution.status = "completed"
            execution.execution_time_seconds = int(time.time() - start_time)
            
            policy.last_executed = datetime.utcnow()
            policy.next_execution = datetime.utcnow() + timedelta(days=1)
            
        except Exception as e:
            execution.status = "failed"
            execution.error_message = str(e)
            execution.execution_time_seconds = int(time.time() - start_time)
        
        self.db.commit()
        self.db.refresh(execution)
        
        return execution
    
    def execute_all_policies(self) -> List[RetentionExecution]:
        """Execute All Policies."""
        """Execute all active retention policies that are due."""
        due_policies = self.db.query(DataRetentionPolicy).filter(
            and_(
                DataRetentionPolicy.status == RetentionStatus.ACTIVE,
                DataRetentionPolicy.next_execution <= datetime.utcnow()
            )
        ).all()
        
        executions = []
        for policy in due_policies:
            try:
                execution = self.execute_policy(policy.id)
                executions.append(execution)
            except Exception as e:
                print(f"Failed to execute policy {policy.policy_code}: {e}")
        
        return executions
    
    def get_policy(self, policy_id: UUID) -> Optional[DataRetentionPolicy]:
        """Get Policy."""
        """Get a retention policy by ID."""
        return self.db.query(DataRetentionPolicy).filter(
            DataRetentionPolicy.id == policy_id
        ).first()
    
    def list_policies(self, active_only: bool = True) -> List[DataRetentionPolicy]:
        """List Policies."""
        """List retention policies."""
        query = self.db.query(DataRetentionPolicy)
        
        if active_only:
            query = query.filter(DataRetentionPolicy.status == RetentionStatus.ACTIVE)
        
        return query.order_by(DataRetentionPolicy.policy_name).all()
    
    def get_execution_history(self, policy_id: Optional[UUID] = None, limit: int = 100) -> List[RetentionExecution]:
        """Get Execution History."""
        """Get retention execution history."""
        query = self.db.query(RetentionExecution)
        
        if policy_id:
            query = query.filter(RetentionExecution.policy_id == policy_id)
        
        return query.order_by(desc(RetentionExecution.execution_date)).limit(limit).all()
    
    def initialize_default_policies(self):
        """Initialize Default Policies."""
        """Initialize default retention policies."""
        default_policies = [
            {
                'policy_name': 'Audit Log Retention',
                'policy_code': 'AUDIT_LOG_7Y',
                'table_name': 'audit_logs',
                'data_category': 'audit',
                'retention_period_days': 2555,
                'retention_action': RetentionAction.ARCHIVE,
                'description': 'Retain audit logs for 7 years as required by regulations',
                'legal_basis': 'SOX compliance requirement'
            },
            {
                'policy_name': 'Session Data Cleanup',
                'policy_code': 'SESSION_30D',
                'table_name': 'user_sessions',
                'data_category': 'session',
                'retention_period_days': 30,
                'retention_action': RetentionAction.DELETE,
                'description': 'Clean up old session data after 30 days',
                'legal_basis': 'Data minimization principle'
            },
            {
                'policy_name': 'Login Attempts Cleanup',
                'policy_code': 'LOGIN_90D',
                'table_name': 'login_attempts',
                'data_category': 'security',
                'retention_period_days': 90,
                'retention_action': RetentionAction.DELETE,
                'description': 'Clean up login attempt records after 90 days',
                'legal_basis': 'Security monitoring requirement'
            }
        ]
        
        for policy_data in default_policies:
            existing = self.db.query(DataRetentionPolicy).filter(
                DataRetentionPolicy.policy_code == policy_data['policy_code']
            ).first()
            
            if not existing:
                policy = DataRetentionPolicy(**policy_data)
                policy.next_execution = datetime.utcnow() + timedelta(days=1)
                self.db.add(policy)
        
        self.db.commit()
    
    def _delete_records(self, policy: DataRetentionPolicy, cutoff_date: datetime) -> Dict[str, int]:
        """ Delete Records."""
        """Delete records based on retention policy."""
        where_clause = f"created_at < :cutoff_date"
        
        if policy.conditions:
            for key, value in policy.conditions.items():
                where_clause += f" AND {key} = :{key}"
        
        count_query = f"SELECT COUNT(*) FROM {policy.table_name} WHERE {where_clause}"
        params = {'cutoff_date': cutoff_date}
        if policy.conditions:
            params.update(policy.conditions)
        
        count_result = self.db.execute(text(count_query), params)
        record_count = count_result.scalar()
        
        delete_query = f"DELETE FROM {policy.table_name} WHERE {where_clause}"
        self.db.execute(text(delete_query), params)
        self.db.commit()
        
        return {
            'processed': record_count,
            'deleted': record_count,
            'archived': 0,
            'anonymized': 0
        }
    
    def _archive_records(self, policy: DataRetentionPolicy, cutoff_date: datetime) -> Dict[str, int]:
        """ Archive Records."""
        """Archive records based on retention policy."""
        where_clause = f"created_at < :cutoff_date"
        
        if policy.conditions:
            for key, value in policy.conditions.items():
                where_clause += f" AND {key} = :{key}"
        
        count_query = f"SELECT COUNT(*) FROM {policy.table_name} WHERE {where_clause}"
        params = {'cutoff_date': cutoff_date}
        if policy.conditions:
            params.update(policy.conditions)
        
        count_result = self.db.execute(text(count_query), params)
        record_count = count_result.scalar()
        
        return {
            'processed': record_count,
            'deleted': 0,
            'archived': record_count,
            'anonymized': 0
        }
    
    def _anonymize_records(self, policy: DataRetentionPolicy, cutoff_date: datetime) -> Dict[str, int]:
        """ Anonymize Records."""
        """Anonymize records based on retention policy."""
        where_clause = f"created_at < :cutoff_date"
        
        if policy.conditions:
            for key, value in policy.conditions.items():
                where_clause += f" AND {key} = :{key}"
        
        count_query = f"SELECT COUNT(*) FROM {policy.table_name} WHERE {where_clause}"
        params = {'cutoff_date': cutoff_date}
        if policy.conditions:
            params.update(policy.conditions)
        
        count_result = self.db.execute(text(count_query), params)
        record_count = count_result.scalar()
        
        return {
            'processed': record_count,
            'deleted': 0,
            'archived': 0,
            'anonymized': record_count
        }